### PR TITLE
fix: DEV-3350: Video object boundaries

### PR DESCRIPTION
**Pull request by nicholasrq**

### CODE REVIEW

File: src/tags/object/Video/Rectangle.tsx [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    const node = e.target;
+
+    if (e.type === 'dragmove') onDragMove(e as KonvaEventObject<DragEvent>);
+
+    reg.updateShape(getNodeAbsoluteDimensions(node, workingArea), frame);
+  };
+
+  const onTransform = (e: KonvaEventObject<Event>) => {
+    const node = e.target;
+    // limitTransformations(e.target);
```
Comment (bmartel) [inline src/tags/object/Video/Rectangle.tsx:55]:
```suggestion
```

File: src/tags/object/Video/Rectangle.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      strokeScaleEnabled={false}
+      selected={selected}
+      draggable={draggable}
+      listening={listening}
+      opacity={reg.hidden ? 0 : 1}
+      onTransform={onTransform}
+      onTransformEnd={onDimensionUpdate}
+      onDragMove={onDimensionUpdate}
+      onDragEnd={onDimensionUpdate}
+      // dragBoundFunc={dragBoundFunc}
```
Comment (bmartel) [inline src/tags/object/Video/Rectangle.tsx:74]:
```suggestion
```

File: src/tags/object/Video/VideoRegions.js [hunk 3]
```diff
@@ -201,11 +208,76 @@ const VideoRegionsPure = ({
...
+}) => {
+  const box = reg.getShape(frame);
+
+  return reg.isInLifespan(frame) && box && (
+    <Rectangle
+      reg={reg}
+      box={box}
+      frame={frame}
+      onClick={(e) => {
+        // if (!reg.annotation.editable || reg.parent.getSkipInteractions()) return;
```
Comment (bmartel) [inline src/tags/object/Video/VideoRegions.js:268]:
```suggestion
```

File: src/tags/object/Video/VideoRegions.js [hunk 4]
```diff
@@ -134,7 +153,7 @@ const VideoRegionsPure = ({
   const handleMouseUp = e => {
     if (!isDrawing) return false;
 
-    const { x, y } = normalizeMouseOffsets(e.evt.offsetX, e.evt.offsetY);
+    const { x, y } = limitCoordinates(normalizeMouseOffsets(e.evt.offsetX, e.evt.offsetY));
     // const { offsetX: x, offsetY: y } = e.evt;
```
Comment (bmartel) [inline src/tags/object/Video/VideoRegions.js:157]:
```suggestion
```

File: src/tags/object/Video/HtxVideo.js [hunk 5]
```diff
@@ -19,6 +19,7 @@ import { VideoRegions } from "./VideoRegions";
 import ResizeObserver from "../../../utils/resize-observer";
 import { useFullscreen } from "../../../hooks/useFullscreen";
 import { IconZoomIn } from "../../../assets/icons";
+import { FF_DEV_3350, isFF } from "../../../utils/feature-flags";
 
 // const hotkeys = Hotkey("Video", "Video Annotation");
```
Comment (bmartel) [inline src/tags/object/Video/HtxVideo.js:24]:
```suggestion
```

File: src/components/ImageView/ImageView.js [hunk 6]
```diff
@@ -505,7 +505,7 @@ export default observer(
 
     handleMouseDown = e => {
       const { item } = this.props;
-  
+
       item.updateSkipInteractions(e);
 
       // item.freezeHistory();
```
Comment (bmartel) [inline src/components/ImageView/ImageView.js:511]:
```suggestion
```

**Review by bmartel — CHANGES_REQUESTED** [review]
The code looks good, can we clean up all these commented out pieces of code? I am going to take another pass over this again as well just as its quite dense.

**Review by bmartel — APPROVED** [review]

File: src/tags/object/Video/tools.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+};
+
+export const normalizeNodeDimentions = <T extends KonvaNode>(node: T, shapeType: "rect") => {
+  const scaleX = node.scaleX();
+  const scaleY = node.scaleY();
+
+  switch (shapeType) {
+    case "rect": {
+      node.width(Math.max(3, node.width() * scaleX));
+      node.height(Math.max(3, node.height() * scaleY));
```
Comment (hlomzik) [inline src/tags/object/Video/tools.ts:25]: magical number. we already have constants for minimal rect size

File: src/tags/object/Video/TransformTools.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+  // if new bounding box is out of visible viewport, let's just skip transforming
+  // this logic can be improved by still allow some transforming if we have small available space
+  if (isOut) {
+    return oldBox;
+  }
+  return newBox;
+};
+
+export const onDragMove = (workingArea: WorkingArea, enabled = true) => function(this: Transformer, e: KonvaEventObject<DragEvent>) {
```
Comment (hlomzik) [inline src/tags/object/Video/TransformTools.ts:86]: it returns function, so name should reflect this — like `createOnDragMoveHandler` or maybe something shorter

File: src/tags/object/Video/TransformTools.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+  return {
+    x: minX,
+    y: minY,
+    width: maxX - minX,
+    height: maxY - minY,
+  };
+};
+
+export const boundBoxFunction = (workingArea: WorkingArea, enabled = true) => (oldBox: Box, newBox: Box) => {
```
Comment (hlomzik) [inline src/tags/object/Video/TransformTools.ts:68]: and that's opposite — it returns bbox, so just `getBBox` or smth

File: src/tags/object/Video/TransformTools.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+  return {
+    x: minX,
+    y: minY,
+    width: maxX - minX,
+    height: maxY - minY,
+  };
+};
+
+export const getTotalBox = (boxes: {
```
Comment (hlomzik) [inline src/tags/object/Video/TransformTools.ts:42]: whole? but doesn't matter

File: src/utils/feature-flags.ts [hunk 11]
```diff
@@ -101,10 +101,21 @@ export const FF_DEV_2461 = "fflag_feat_front_dev_2461_audio_paragraphs_seek_chun
  */
 export const FF_DEV_2918 = "fflag_fix_front_dev_2918_labeling_filtered_paragraphs_250822_short";
 
+/**
+ * Disallow drawing regions outside of the video canvas
+ * Also disables offscreen zooming and panning
+ * @link https://app.launchdarkly.com/default/production/features/fflag_fix_front_dev_3350_restrict_drawing_area_short
```
Comment (hlomzik) [inline src/utils/feature-flags.ts:107]: Not sure it's a good idea to hardcode such links, especially in opensource. This page can be easily found by ff name and the whole comment is too big. I'd suggest to use simple one-line comment.

**Review by hlomzik — CHANGES_REQUESTED** [review]

**Review by hlomzik — APPROVED** [review]

File: public/index.html [hunk 12]
```diff
@@ -95,7 +95,7 @@
               "instruction",
               "side-column",
               "ground-truth",
-              "annotations:history",
+              //"annotations:history",
```
[the remark this answers is not in the document]
Comment (yyassi-heartex) [reply public/index.html:98]:
```suggestion
```

**Review by yyassi-heartex — COMMENTED** [review]
really minor tweaks, removing the commented out code + reusing the min rectangle size const as @hlomzik suggested then we should be good to go

**Review by yyassi-heartex — APPROVED** [review]

**Pull request merged**